File: src/test/java/com/d20charactersheet/framework/boc/service/DnDv35RuleServiceSavingThrowTest.java
package com.d20charactersheet.framework.boc.service;

import static org.junit.Assert.assertEquals;

import java.util.ArrayList;
import java.util.List;

import org.junit.Test;

import com.d20charactersheet.framework.boc.model.Character;
import com.d20charactersheet.framework.boc.model.ClassLevel;
import com.d20charactersheet.framework.boc.model.Save;

public class DnDv35RuleServiceSavingThrowTest extends DnDv35RuleServiceTestCase {

  @Test
  public void testGetGoodSavingThrow() {
    final AbstractRuleServiceImpl abstractRuleServiceImpl = (AbstractRuleServiceImpl) ruleService;

    // level 1 => +2
    int savingThrow = abstractRuleServiceImpl.getGoodSave(1);
    assertEquals(2, savingThrow);

    // level 2 => +3
    savingThrow = abstractRuleServiceImpl.getGoodSave(2);
    assertEquals(3, savingThrow);

    // level 3 => +3
    savingThrow = abstractRuleServiceImpl.getGoodSave(3);
    assertEquals(3, savingThrow);

    // level 10 => +7
    savingThrow = abstractRuleServiceImpl.getGoodSave(10);
    assertEquals(7, savingThrow);

    // level 15 => +9
    savingThrow = abstractRuleServiceImpl.getGoodSave(15);
    assertEquals(9, savingThrow);

    // level 20 => +12
    savingThrow = abstractRuleServiceImpl.getGoodSave(20);
    assertEquals(12, savingThrow);
  }

  @Test
  public void testGetPoorSavingThrow() {
    final AbstractRuleServiceImpl abstractRuleServiceImpl = (AbstractRuleServiceImpl) ruleService;

    // level 1 => 0
    int savingThrow = abstractRuleServiceImpl.getPoorSave(1);
    assertEquals(0, savingThrow);

    // level 2 => 0
    savingThrow = abstractRuleServiceImpl.getPoorSave(2);
    assertEquals(0, savingThrow);

    // level 3 => +1
    savingThrow = abstractRuleServiceImpl.getPoorSave(3);
    assertEquals(1, savingThrow);

    // level 10 => +3
    savingThrow = abstractRuleServiceImpl.getPoorSave(10);
    assertEquals(3, savingThrow);

    // level 15 => +5
    savingThrow = abstractRuleServiceImpl.getPoorSave(15);
    assertEquals(5, savingThrow);

    // level 20 => +6
    savingThrow = abstractRuleServiceImpl.getPoorSave(20);
    assertEquals(6, savingThrow);
  }

  @Test
  public void testGetSavingThrowClass() {
    final AbstractRuleServiceImpl abstractRuleServiceImpl = (AbstractRuleServiceImpl) ruleService;

    // Fighter (1), Fortitude => 2
    int savingThrow = abstractRuleServiceImpl.getBaseSave(new ClassLevel(fighter, 1), Save.FORTITUDE);
    assertEquals(2, savingThrow);

    // Fighter (1), Reflex => 0
    savingThrow = abstractRuleServiceImpl.getBaseSave(new ClassLevel(fighter, 1), Save.REFLEX);
    assertEquals(0, savingThrow);

    // Fighter (1), Will => 0
    savingThrow = abstractRuleServiceImpl.getBaseSave(new ClassLevel(fighter, 1), Save.WILL);
    assertEquals(0, savingThrow);

    // Fighter (3), Fortitude => 3
    savingThrow = abstractRuleServiceImpl.getBaseSave(new ClassLevel(fighter, 3), Save.FORTITUDE);
    assertEquals(3, savingThrow);

    // Fighter (3), Reflex => 1
    savingThrow = abstractRuleServiceImpl.getBaseSave(new ClassLevel(fighter, 3), Save.REFLEX);
    assertEquals(1, savingThrow);

  }

  @Test
  public void testGetSavingThrowCharacter() {
    final Character character = createCharacter(wizard, 5);
    character.setConstitution(12);
    character.setDexterity(14);
    character.setWisdom(16);
    character.setFortitudeModifier(1);
    character.setReflexModifier(2);
    character.setWillModifier(3);

    int savingThrow = ruleService.getSave(character, Save.FORTITUDE);
    assertEquals(3, savingThrow);

    savingThrow = ruleService.getSave(character, Save.REFLEX);
    assertEquals(5, savingThrow);

    savingThrow = ruleService.getSave(character, Save.WILL);
    assertEquals(10, savingThrow);
  }

  @Test
  public void testGetSavingThrowMultiClassCharacter() {
    final Character character = new Character();
    final List<ClassLevel> classLevels = new ArrayList<>();
    classLevels.add(new ClassLevel(ranger, 4));
    classLevels.add(new ClassLevel(rogue, 3));
    character.setClassLevels(classLevels);
    character.setConstitution(12);
    character.setDexterity(14);
    character.setWisdom(16);
    character.setFortitudeModifier(1);
    character.setReflexModifier(2);
    character.setWillModifier(3);

    // Ranger(4) + Rogue(3) + Constitution + modifier = +4 +1 +1 +1 = 7
    int savingThrow = ruleService.getSave(character, Save.FORTITUDE);
    assertEquals(7, savingThrow);

    // Ranger(4) + Rogue(3) + Constitution + modifier = +4 +3 +2 +2 = 11
    savingThrow = ruleService.getSave(character, Save.REFLEX);
    assertEquals(11, savingThrow);

    // Ranger(4) + Rogue(3) + Constitution = +1 +1 +3 +3 = 8
    savingThrow = ruleService.getSave(character, Save.WILL);
    assertEquals(8, savingThrow);

  }

  @Test
  public void testGetBaseSaveMultiCharacter() {
    final Character character = new Character();
    final List<ClassLevel> classLevels = new ArrayList<>();
    classLevels.add(new ClassLevel(ranger, 4));
    classLevels.add(new ClassLevel(rogue, 3));
    character.setClassLevels(classLevels);

    // Ranger(4) + Rogue(3) = +4 +1 = 5
    int savingThrow = ruleService.getBaseSave(character, Save.FORTITUDE);
    assertEquals(5, savingThrow);

    // Ranger(4) + Rogue(3) = +4 +3 = 7
    savingThrow = ruleService.getBaseSave(character, Save.REFLEX);
    assertEquals(7, savingThrow);

    // Ranger(4) + Rogue(3) = +1 +1 = 2
    savingThrow = ruleService.getBaseSave(character, Save.WILL);
    assertEquals(2, savingThrow);

  }

}

File: src/test/java/com/d20charactersheet/framework/boc/service/DnDv35RuleServiceSavingThrowTest.kt
package com.d20charactersheet.framework.boc.service

import com.d20charactersheet.framework.boc.model.Character
import com.d20charactersheet.framework.boc.model.CharacterClass
import com.d20charactersheet.framework.boc.model.ClassLevel
import com.d20charactersheet.framework.boc.model.Save
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.util.*

class DnDv35RuleServiceSavingThrowTest {

    val ruleService = DnDv35RuleServiceImpl()

    val wizard = CharacterClass().apply { saves = EnumSet.of(Save.CONSTITUTION) }
    val fighter = CharacterClass().apply { saves = EnumSet.of(Save.STRENGTH) }
    val ranger = CharacterClass().apply { saves = EnumSet.of(Save.STRENGTH, Save.DEXTERITY) }
    val rogue = CharacterClass().apply { saves = EnumSet.of(Save.DEXTERITY) }

    @Test
    fun getGoodSave_level1_return2() {
        // act
        val savingThrow = ruleService.getGoodSave(1)

        // assert
        assertThat(savingThrow).isEqualTo(2)
    }

    @Test
    fun getGoodSave_level2_return3() {
        // act
        val savingThrow = ruleService.getGoodSave(2)

        // assert
        assertThat(savingThrow).isEqualTo(3)
    }

    @Test
    fun getGoodSave_level3_return3() {
        // act
        val savingThrow = ruleService.getGoodSave(3)

        // assert
        assertThat(savingThrow).isEqualTo(3)
    }

    @Test
    fun getGoodSave_level10_return7() {
        // act
        val savingThrow = ruleService.getGoodSave(10)

        // assert
        assertThat(savingThrow).isEqualTo(7)
    }

    @Test
    fun getGoodSave_level15_return9() {
        // act
        val savingThrow = ruleService.getGoodSave(15)

        // assert
        assertThat(savingThrow).isEqualTo(9)
    }

    @Test
    fun getGoodSave_level20_return12() {
        // act
        val savingThrow = ruleService.getGoodSave(20)

        // assert
        assertThat(savingThrow).isEqualTo(12)
    }

    @Test
    fun getPoorSave_level1_return0() {
        // act
        val savingThrow = ruleService.getPoorSave(1)

        // assert
        assertThat(savingThrow).isEqualTo(0)
    }

    @Test
    fun getPoorSave_level2_return0() {
        // act
        val savingThrow = ruleService.getPoorSave(2)

        // assert
        assertThat(savingThrow).isEqualTo(0)
    }

    @Test
    fun getPoorSave_level3_return1() {
        // act
        val savingThrow = ruleService.getPoorSave(3)

        // assert
        assertThat(savingThrow).isEqualTo(1)
    }

    @Test
    fun getPoorSave_level10_return3() {
        // act
        val savingThrow = ruleService.getPoorSave(10)

        // assert
        assertThat(savingThrow).isEqualTo(3)
    }

    @Test
    fun getPoorSave_level15_return5() {
        // act
        val savingThrow = ruleService.getPoorSave(15)

        // assert
        assertThat(savingThrow).isEqualTo(5)
    }

    @Test
    fun getPoorSave_level20_return6() {
        // act
        val savingThrow = ruleService.getPoorSave(20)

        // assert
        assertThat(savingThrow).isEqualTo(6)
    }

    @Test
    fun getBaseSave_fighterLevel1_fortitudeIs2() {
        // act
        val savingThrow = ruleService.getBaseSave(ClassLevel(fighter, 1), Save.STRENGTH)

        // assert
        assertThat(savingThrow).isEqualTo(2)
    }

    @Test
    fun getBaseSave_fighterLevel1_reflexIs0() {
        // act
        val savingThrow = ruleService.getBaseSave(ClassLevel(fighter, 1), Save.DEXTERITY)

        // assert
        assertThat(savingThrow).isEqualTo(0)
    }

    @Test
    fun getBaseSave_fighterLevel1_willIs0() {
        // act
        val savingThrow = ruleService.getBaseSave(ClassLevel(fighter, 1), Save.CONSTITUTION)

        // assert
        assertThat(savingThrow).isEqualTo(0)
    }

    @Test
    fun getBaseSave_fighterLevel3_fortitudeIs3() {
        // act
        val savingThrow = ruleService.getBaseSave(ClassLevel(fighter, 3), Save.STRENGTH)

        // assert
        assertThat(savingThrow).isEqualTo(3)
    }

    @Test
    fun getBaseSave_fighterLevel3_reflexIs1() {
        // act
        val savingThrow = ruleService.getBaseSave(ClassLevel(fighter, 3), Save.DEXTERITY)

        // assert
        assertThat(savingThrow).isEqualTo(1)
    }

    @Test
    fun getSave_singleClassWizardLevel5_fortitudeIs3() {
        // arrange
        val character = Character().apply {
            classLevels = listOf(
                ClassLevel(wizard, 5)
            )
        }
        character.constitution = 12
        character.dexterity = 14
        character.wisdom = 16
        character.setSaveModifier(Save.STRENGTH, 1)
        character.setSaveModifier(Save.DEXTERITY, 2)
        character.setSaveModifier(Save.CONSTITUTION, 3)

        // act
        val savingThrow = ruleService.getSave(character, Save.STRENGTH)

        // assert
        assertThat(savingThrow).isEqualTo(3)
    }

    @Test
    fun getSave_singleClassWizardLevel5_reflexIs5() {
        // arrange
        val character = Character().apply {
            classLevels = listOf(
                ClassLevel(wizard, 5)
            )
        }
        character.constitution = 12
        character.dexterity = 14
        character.wisdom = 16
        character.setSaveModifier(Save.STRENGTH, 1)
        character.setSaveModifier(Save.DEXTERITY, 2)
        character.setSaveModifier(Save.CONSTITUTION, 3)

        // act
        val savingThrow = ruleService.getSave(character, Save.DEXTERITY)

        // assert
        assertThat(savingThrow).isEqualTo(5)
    }

    @Test
    fun getSave_singleClassWizardLevel5_willIs10() {
        // arrange
        val character = Character().apply {
            classLevels = listOf(
                ClassLevel(wizard, 5)
            )
        }
        character.constitution = 12
        character.dexterity = 14
        character.wisdom = 16
        character.setSaveModifier(Save.STRENGTH, 1)
        character.setSaveModifier(Save.DEXTERITY, 2)
        character.setSaveModifier(Save.CONSTITUTION, 3)

        // act
        val savingThrow = ruleService.getSave(character, Save.CONSTITUTION)

        // assert
        assertThat(savingThrow).isEqualTo(10)
    }

    @Test
    fun getSave_ranger4Rogu3_fortitude7() {

        // arrange
        val character = Character()
        val classLevels: MutableList<ClassLevel> = ArrayList()
        classLevels.add(ClassLevel(ranger, 4))
        classLevels.add(ClassLevel(rogue, 3))
        character.classLevels = classLevels
        character.constitution = 12
        character.dexterity = 14
        character.wisdom = 16
        character.setSaveModifier(Save.STRENGTH, 1)
        character.setSaveModifier(Save.DEXTERITY, 2)
        character.setSaveModifier(Save.CONSTITUTION, 3)

        // act
        val savingThrow = ruleService.getSave(character, Save.STRENGTH)

        // assert
        // Ranger(4) + Rogue(3) + Constitution + modifier = +4 +1 +1 +1 = 7
        assertThat(savingThrow).isEqualTo(7)
    }

    @Test
    fun getSave_ranger4Rogu3_reflex111() {

        // arrange
        val character = Character()
        val classLevels: MutableList<ClassLevel> = ArrayList()
        classLevels.add(ClassLevel(ranger, 4))
        classLevels.add(ClassLevel(rogue, 3))
        character.classLevels = classLevels
        character.constitution = 12
        character.dexterity = 14
        character.wisdom = 16
        character.setSaveModifier(Save.STRENGTH, 1)
        character.setSaveModifier(Save.DEXTERITY, 2)
        character.setSaveModifier(Save.CONSTITUTION, 3)

        // act
        val savingThrow = ruleService.getSave(character, Save.DEXTERITY)

        // assert
        // Ranger(4) + Rogue(3) + Constitution + modifier = +4 +3 +2 +2 = 11
        assertThat(savingThrow).isEqualTo(11)
    }

    @Test
    fun getSave_ranger4Rogu3_will8() {

        // arrange
        val character = Character()
        val classLevels: MutableList<ClassLevel> = ArrayList()
        classLevels.add(ClassLevel(ranger, 4))
        classLevels.add(ClassLevel(rogue, 3))
        character.classLevels = classLevels
        character.constitution = 12
        character.dexterity = 14
        character.wisdom = 16
        character.setSaveModifier(Save.STRENGTH, 1)
        character.setSaveModifier(Save.DEXTERITY, 2)
        character.setSaveModifier(Save.CONSTITUTION, 3)

        // act
        val savingThrow = ruleService.getSave(character, Save.CONSTITUTION)

        // Ranger(4) + Rogue(3) + Constitution = +1 +1 +3 +3 = 8
        assertThat(savingThrow).isEqualTo(8)
    }


    @Test
    fun getBase_ranger4rogue3_fortitude() {

        // arrange
        val character = Character()
        val classLevels: MutableList<ClassLevel> = ArrayList()
        classLevels.add(ClassLevel(ranger, 4))
        classLevels.add(ClassLevel(rogue, 3))
        character.classLevels = classLevels

        // act
        val savingThrow = ruleService.getProficiencySave(character, Save.STRENGTH)

        // assert
        // Ranger(4) + Rogue(3) = +4 +1 = 5
        assertThat(savingThrow).isEqualTo(5)
    }

    @Test
    fun getBase_ranger4rogue3_reflex7() {

        // arrange
        val character = Character()
        val classLevels: MutableList<ClassLevel> = ArrayList()
        classLevels.add(ClassLevel(ranger, 4))
        classLevels.add(ClassLevel(rogue, 3))
        character.classLevels = classLevels

        // act
        val savingThrow = ruleService.getProficiencySave(character, Save.DEXTERITY)

        // assert
        // Ranger(4) + Rogue(3) = +4 +3 = 7
        assertThat(savingThrow).isEqualTo(7)
    }

    @Test
    fun getBase_ranger4rogue3_will2() {

        // arrange
        val character = Character()
        val classLevels: MutableList<ClassLevel> = ArrayList()
        classLevels.add(ClassLevel(ranger, 4))
        classLevels.add(ClassLevel(rogue, 3))
        character.classLevels = classLevels

        // act
        val savingThrow = ruleService.getProficiencySave(character, Save.CONSTITUTION)

        // assert
        // Ranger(4) + Rogue(3) = +1 +1 = 2
        assertThat(savingThrow).isEqualTo(2)
    }
}